compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -827,19 +829,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 if expected_ty == self.tcx.types.bool {
                                     // If this is caused by a missing `let` in a `while let`,
                                     // silence this redundant error, as we already emit E0070.
-                                    let parent = self.tcx.hir().get_parent_node(blk.hir_id);
-                                    let parent = self.tcx.hir().get_parent_node(parent);
-                                    let parent = self.tcx.hir().get_parent_node(parent);
-                                    let parent = self.tcx.hir().get_parent_node(parent);
-                                    let parent = self.tcx.hir().get_parent_node(parent);
-                                    match self.tcx.hir().find(parent) {
-                                        Some(hir::Node::Expr(hir::Expr {
-                                            kind: hir::ExprKind::Loop(_, _, hir::LoopSource::While, _),
-                                            ..
-                                        })) => {
+
+                                    // Our block must be a `assign desugar local; assignment`
+                                    if let Some(hir::Node::Block(hir::Block {
+                                        stmts:
+                                            [hir::Stmt {
+                                                kind:
+                                                    hir::StmtKind::Local(hir::Local {
+                                                        source: hir::LocalSource::AssignDesugar(_),
+                                                        ..
+                                                    }),
+                                                ..
+                                            }, hir::Stmt {
+                                                kind:
+                                                    hir::StmtKind::Expr(hir::Expr {
+                                                        kind: hir::ExprKind::Assign(..),
+                                                        ..
+                                                    }),
+                                                ..
+                                            }],
+                                        ..
+                                    })) = self.tcx.hir().find(blk.hir_id)
+                                    {
+                                        self.comes_from_where_condition(blk.hir_id, |_| {
                                             err.delay_as_bug();
```

Comment:
This is going to make @eddyb unhappy 😬, given their PR removing them 😅

Reply:
Sorry, removing what? `delay_as_bug`?

Reply:
Yes

Reply:
I don't remove `delay_as_bug` - not only is it kept, but #93368 makes it available on `Diagnostic` itself (as `.downgrade_to_delayed_bug()`, which will cause the later emission of the `Diagnostic` to be equivalent to `.delay_as_bug()`).

Reply:
Then I might go ahead and continue with this approach, and wait until #93368 and use `downgrade_to_delayed_bug` instead? :)